x/ibc/07-tendermint/misbehaviour.go

```diff
@@ -0,0 +1,90 @@
+package tendermint
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clientexported "github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	ibctypes "github.com/cosmos/cosmos-sdk/x/ibc/types"
+)
+
+// CheckMisbehaviourAndUpdateState determines whether or not two conflicting
+// headers at the same height would have convinced the light client.
+//
+// NOTE: assumes provided height is the height at which the consensusState is
+// stored.
+func CheckMisbehaviourAndUpdateState(
+	clientState clientexported.ClientState,
+	consensusState clientexported.ConsensusState,
+	misbehaviour clientexported.Misbehaviour,
+	height uint64, // height at which the consensus state was loaded
+) (clientexported.ClientState, error) {
+
+	// cast the interface to specific types before checking for misbehaviour
+	tmClientState, ok := clientState.(ClientState)
+	if !ok {
+		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidClientType, "client state type is not Tendermint")
+	}
+
+	tmConsensusState, ok := consensusState.(ConsensusState)
+	if !ok {
+		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidClientType, "consensus state is not Tendermint")
+	}
+
+	tmEvidence, ok := misbehaviour.(Evidence)
+	if !ok {
+		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidClientType, "evidence type is not Tendermint")
+	}
+
+	if err := checkMisbehaviour(tmClientState, tmConsensusState, tmEvidence, height); err != nil {
+		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidEvidence, err.Error())
+	}
+
+	tmClientState.FrozenHeight = uint64(tmEvidence.GetHeight())
```

Comment:
In the spec this is set directly to the store, not returned. I think this is the same question as above.

Reply:
Also this calculation is not quite right, it should be the minimum of the height from either of the headers

Reply:
> In the spec this is set directly to the store, not returned. I think this is the same question as above.

blocked on #5502 

> Also this calculation is not quite right, it should be the minimum of the height from either of the headers

makes sense